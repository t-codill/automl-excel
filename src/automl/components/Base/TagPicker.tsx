import { BasePicker, IBasePickerState, IBasePickerStyleProps, IBasePickerStyles, ITag, ITagItemProps, ITagPickerProps, styled, TagItem, TagItemSuggestion } from "office-ui-fabric-react";
// tslint:disable-next-line:no-submodule-imports
import { getStyles } from "office-ui-fabric-react/lib/components/pickers/BasePicker.styles";
import * as React from "react";

export class TagPickerWrapper extends BasePicker<ITag, ITagPickerProps> {
  public static defaultProps = {
    onRenderItem: (props: ITagItemProps) => <TagItem {...props}>{props.item.name}</TagItem>,
    onRenderSuggestionsItem: (props: ITag) => <TagItemSuggestion>{props.name}</TagItemSuggestion>,
  };

  public componentDidUpdate(prevProp: ITagPickerProps, prevState: IBasePickerState): void {
    super.componentDidUpdate(prevProp, prevState);
    if (!this.state.suggestionsVisible && this.state.isFocused) {
      this.onEmptyInputFocus();
      this.setState({ suggestionsVisible: true });
    }
  }
}

// tslint:disable-next-line:variable-name
export const TagPicker = styled<ITagPickerProps, IBasePickerStyleProps, IBasePickerStyles>(TagPickerWrapper, getStyles, undefined, {
  scope: "TagPicker"
});
