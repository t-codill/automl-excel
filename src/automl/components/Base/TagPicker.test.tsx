import { shallow, ShallowWrapper } from "enzyme";
import { BasePicker, IBasePickerState, ITag, ITagItemProps, ITagPickerProps } from "office-ui-fabric-react";
import * as React from "react";
import { TagPickerWrapper } from "./TagPicker";

describe("TagPicker", () => {
    const onRemoveSuggestion = jest.fn();
    function onResolveSuggestions(text: string): ITag[] {
        return [
            "black",
            "blue",
            "brown",
            "cyan",
            "green",
            "magenta",
            "mauve",
            "orange",
            "pink",
            "purple",
            "red",
            "rose",
            "violet",
            "white",
            "yellow"
        ]
            .filter((tag) => tag.toLowerCase()
                .indexOf(text.toLowerCase()) === 0)
            .map((item) => ({ key: item, name: item }));
    }
    const iTagProps: ITagItemProps = {
        className: undefined,
        enableTagFocusInDisabledPicker: true,
        item: {
            key: "Test Key",
            name: "Test Name"
        },
        index: 0
    };
    const iTag: ITag = {
        name: "Test Tag Name",
        key: "Test Tag Key"
    };

    let wrapper: ShallowWrapper<ITagPickerProps, IBasePickerState>;
    const props: ITagPickerProps = {
        onRemoveSuggestion,
        onResolveSuggestions
    };
    beforeEach(() => {

        jest.spyOn(BasePicker.prototype as unknown as { onEmptyInputFocus(): void }, "onEmptyInputFocus")
            .mockImplementation(() => {
                return;
            });
        wrapper = shallow(
            <TagPickerWrapper {...props}
            />);
    });

    it("should render", () => {
        expect(wrapper)
            .toMatchSnapshot();
    });

    it("should render item and suggestion", () => {
        const renderItem = TagPickerWrapper.defaultProps.onRenderItem(iTagProps);
        const renderSuggestionItem = TagPickerWrapper.defaultProps.onRenderSuggestionsItem(iTag);
        expect(renderItem)
            .toMatchSnapshot();
        expect(renderSuggestionItem)
            .toMatchSnapshot();
    });

    it("should trigger update", () => {
        wrapper.setState({
            suggestionsVisible: false,
            isFocused: true
        });
        wrapper.update();
        expect(wrapper)
            .toMatchSnapshot();
    });
});
