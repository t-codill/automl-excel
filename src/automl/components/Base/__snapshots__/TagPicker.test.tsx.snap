// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TagPicker should render 1`] = `
<div
  className="ms-BasePicker"
  onBlur={[Function]}
  onKeyDown={[Function]}
>
  <FocusZone
    componentRef={
      Object {
        "current": null,
      }
    }
    direction={2}
    isCircularNavigation={false}
    isInnerZoneKeystroke={[Function]}
  >
    <SelectionZone
      isMultiSelectEnabled={true}
      isSelectedOnFocus={true}
      selection={
        Selection {
          "_anchoredIndex": 0,
          "_canSelectItem": [Function],
          "_changeEventSuppressionCount": 0,
          "_exemptedCount": 0,
          "_exemptedIndices": Object {},
          "_getKey": [Function],
          "_isModal": false,
          "_items": Array [],
          "_keyToIndexMap": Object {},
          "_onSelectionChanged": [Function],
          "_selectedItems": null,
          "_unselectableCount": 0,
          "_unselectableIndices": Object {},
          "count": 0,
          "mode": 2,
        }
      }
      selectionMode={2}
    >
      <div
        className="ms-BasePicker-text pickerText_de6ce0eb"
      >
        <Autofill
          aria-autocomplete="both"
          aria-controls=" "
          aria-expanded={false}
          aria-haspopup="true"
          autoCapitalize="off"
          autoComplete="off"
          className="ms-BasePicker-input pickerInput_de6ce0eb"
          componentRef={
            Object {
              "current": null,
            }
          }
          enableAutofillOnKeyPress={
            Array [
              40,
              38,
            ]
          }
          onBlur={[Function]}
          onFocus={[Function]}
          onInputValueChange={[Function]}
          role="combobox"
          spellCheck={false}
          suggestedDisplayValue=""
        />
      </div>
    </SelectionZone>
  </FocusZone>
</div>
`;

exports[`TagPicker should render item and suggestion 1`] = `
<Styled
  enableTagFocusInDisabledPicker={true}
  index={0}
  item={
    Object {
      "key": "Test Key",
      "name": "Test Name",
    }
  }
>
  Test Name
</Styled>
`;

exports[`TagPicker should render item and suggestion 2`] = `
<Styled>
  Test Tag Name
</Styled>
`;

exports[`TagPicker should trigger update 1`] = `
<div
  className="ms-BasePicker"
  onBlur={[Function]}
  onKeyDown={[Function]}
>
  <FocusZone
    componentRef={
      Object {
        "current": null,
      }
    }
    direction={2}
    isCircularNavigation={false}
    isInnerZoneKeystroke={[Function]}
  >
    <SelectionZone
      isMultiSelectEnabled={true}
      isSelectedOnFocus={true}
      selection={
        Selection {
          "_anchoredIndex": 0,
          "_canSelectItem": [Function],
          "_changeEventSuppressionCount": 0,
          "_exemptedCount": 0,
          "_exemptedIndices": Object {},
          "_getKey": [Function],
          "_isModal": false,
          "_items": Array [],
          "_keyToIndexMap": Object {},
          "_onSelectionChanged": [Function],
          "_selectedItems": null,
          "_unselectableCount": 0,
          "_unselectableIndices": Object {},
          "count": 0,
          "mode": 2,
        }
      }
      selectionMode={2}
    >
      <div
        className="ms-BasePicker-text pickerText_de6ce0eb inputFocused_de6ce0eb"
      >
        <Autofill
          aria-autocomplete="both"
          aria-controls="suggestion-list-id__3 "
          aria-expanded={true}
          aria-haspopup="true"
          aria-owns="suggestion-list-id__3"
          autoCapitalize="off"
          autoComplete="off"
          className="ms-BasePicker-input pickerInput_de6ce0eb"
          componentRef={
            Object {
              "current": null,
            }
          }
          enableAutofillOnKeyPress={
            Array [
              40,
              38,
            ]
          }
          onBlur={[Function]}
          onFocus={[Function]}
          onInputValueChange={[Function]}
          role="combobox"
          spellCheck={false}
          suggestedDisplayValue=""
        />
      </div>
    </SelectionZone>
  </FocusZone>
  <Callout
    directionalHint={4}
    directionalHintForRTL={6}
    gapSpace={5}
    isBeakVisible={false}
    onDismiss={[Function]}
  >
    <StyledSuggestions
      componentRef={
        Object {
          "current": null,
        }
      }
      isMostRecentlyUsedVisible={false}
      isSearching={false}
      moreSuggestionsAvailable={false}
      onGetMoreResults={[Function]}
      onRenderSuggestion={[Function]}
      onSuggestionClick={[Function]}
      onSuggestionRemove={[Function]}
      refocusSuggestions={[Function]}
      suggestions={Array []}
      suggestionsListId="suggestion-list-id__3"
    />
  </Callout>
</div>
`;
